src/psyclone/parse/module_info.py

```diff
@@ -0,0 +1,213 @@
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author J. Henrichs, Bureau of Meteorology
+
+'''This module contains the ModuleInfo class, which is used to store
+and cache information about a module: the filename, source code (if requested)
+and the fparser tree (if requested).
+'''
+
+
+from fparser.common.readfortran import FortranStringReader
+from fparser.two.Fortran2003 import Use_Stmt
+from fparser.two.parser import ParserFactory
+from fparser.two.utils import walk
+
+from psyclone.errors import PSycloneError
+
+
+# ============================================================================
+class ModuleInfoError(PSycloneError):
+    '''
+    PSyclone-specific exception for use when an error with the module manager
+    happens - typically indicating that some module information cannot be
+    found.
+
+    :param str value: the message associated with the error.
+
+    '''
+    def __init__(self, value):
+        PSycloneError.__init__(self, value)
+        self.value = "ModuleInfo error: "+str(value)
+
+
+# ============================================================================
+class ModuleInfo:
+    '''This class stores mostly cached information about modules: it stores
+    the original filename, if requested it will read the file and then caches
+    the plain text file, and if required it will parse the file, and then
+    cache the fparser AST.
+
+    :param str name: the module name.
+    :param str filename: the name of the source file that stores this module \
+        (including path).
+
+    '''
+
+    def __init__(self, name, filename):
+        self._name = name
+        self._filename = filename
+        # A cache for the source code:
+        self._source_code = None
+
+        # A cache for the fparser tree
+        self._parse_tree = None
+
+        # A cache for the module dependencies: this is just a set
+        # of all modules used by this module. Type: Set[str]
+        self._used_modules = None
+
+        # This is a dictionary, with the modules from the list of all used
```

Comment:
"...dictionary containing the sets of symbols imported from each module, indexed by the module names: Dict[str, Set(str)]"

Reply:
Done.